tests/test_templates_utils.py

```diff
@@ -0,0 +1,209 @@
+            _check_shape(inpt, target_shape, bound=bound)
+
+    @pytest.mark.parametrize("inpt, target_shape, bound", SHAPE_LST_FAIL)
+    def test_check_shape_list_of_inputs_exception(self, inpt, target_shape, bound):
+        with pytest.raises(ValueError, match="Input has shape"):
+            _check_shapes(inpt, target_shape, bound_list=[bound]*len(inpt))
+
+    @pytest.mark.parametrize("inpt, target_shape", GET_SHAPE_PASS)
+    def test_get_shape(self, inpt, target_shape):
+        shape = _get_shape(inpt)
+        assert shape == target_shape
+
+    @pytest.mark.parametrize("inpt", GET_SHAPE_FAIL)
+    def test_get_shape_exception(self, inpt):
+        with pytest.raises(ValueError):
+            _get_shape(inpt)
+
+    @pytest.mark.parametrize("inpt, repeat", LAYERS_PASS)
+    def test_check_num_layers(self, inpt, repeat):
+        n_layers = _check_number_of_layers(inpt)
+        assert n_layers == repeat
+
+    @pytest.mark.parametrize("inpt, repeat", LAYERS_FAIL)
+    def test_check_num_layers_exception(self, inpt, repeat):
+        with pytest.raises(ValueError, match="The first dimension of the weight parameters"):
+            _check_number_of_layers(inpt)
+
+    def test_check_shape_exception_message(self):
+        with pytest.raises(ValueError, match="XXX"):
+            _check_shape([0.], (3,), msg="XXX")
+
+    @pytest.mark.parametrize("arg", NO_VARIABLES_PASS)
+    def test_check_no_variables(self, arg):
+        _check_no_variable(arg, "dummy")
+
+    @pytest.mark.parametrize("arg", NO_VARIABLES_FAIL)
+    def test_check_no_variables_exception(self, arg):
+        with pytest.raises(ValueError, match="The argument dummy can not be passed"):
+            _check_no_variable(arg, "dummy")
+
+    def test_check_no_variables_exception_message(self):
+        with pytest.raises(ValueError, match="XXX"):
+            a = Variable(0)
+            _check_no_variable([a], ["dummy"], msg="XXX")
+
+    @pytest.mark.parametrize("hp, opts", OPTIONS_PASS)
+    def test_check_hyperp_options(self, hp, opts):
+        _check_hyperp_is_in_options(hp, opts)
+
+    @pytest.mark.parametrize("hp, opts", OPTIONS_FAIL)
+    def test_check_hyperp_options_exception(self, hp, opts):
+        with pytest.raises(ValueError, match="Hyperparameter c must be one"):
+            _check_hyperp_is_in_options(hp, opts)
+
+    @pytest.mark.parametrize("hp, typ, alt", TYPE_PASS)
+    def test_check_type(self, hp, typ, alt):
+        _check_type(hp, [typ, alt])
+
+    @pytest.mark.parametrize("hp, typ, alt", TYPE_FAIL)
+    def test_check_type_exception(self, hp, typ, alt):
+        with pytest.raises(ValueError, match="Hyperparameter type must be one"):
+            _check_type(hp, [typ, alt])
+
+    @pytest.mark.parametrize("hp, typ, alt", TYPE_FAIL)
+    def test_check_type_exception_message(self, hp, typ, alt):
+        with pytest.raises(ValueError, match="XXX"):
+            _check_type(hp, [typ, alt], msg="XXX")
```

Comment:
TODO: These tests should have docstrings

Reply:
Done...